### PR TITLE
Add convenience functions for loading the index

### PR DESCRIPTION
**Pull request by sgallagher**
Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

Attn: @FrostyX

This would replace the need for a couple of the proposed `modulemd_tools` functions you're writing.

### CODE REVIEW

**Comment by FrostyX** [comment]
Thank you @sgallagher,
those functions are surely helpful, but personally I would find a better use of a function that takes a file or a string containing the YAML content and returning `Modulemd.ModuleStreamV2`. Basically I would like to take a YAML and being able to access the module attributes (e.g. name, stream, summary, ...) in the least possible steps.

**Review by mmathesius — CHANGES_REQUESTED** [review]
What's here looks good, but a couple simple tests would be in order. It looks like `modulemd_get_version()` is missing any tests, too.

**Comment by mmathesius** [comment]
> Thank you @sgallagher,
> those functions are surely helpful, but personally I would find a better use of a function that takes a file or a string containing the YAML content and returning `Modulemd.ModuleStreamV2`. Basically I would like to take a YAML and being able to access the module attributes (e.g. name, stream, summary, ...) in the least possible steps.

I'm probably missing some context, but don't `modulemd_module_stream_read_file()` and `modulemd_module_stream_read_string()` provide that?

**Comment by FrostyX** [comment]
> I'm probably missing some context, but don't `modulemd_module_stream_read_file()` and `modulemd_module_stream_read_string()` provide that?

As far as I understand it, they don't. Those functions return `ModulemdModuleIndex` and there is AFAIK a cumbersome number of steps to actually access the attributes. Maybe I am doing it wrong, but didn't find any simpler way than this:

```
    # idx = Modulemd.ModuleIndex.new()
    # idx.update_from_string(mod_yaml, strict=True)
    idx = ... # provided by some of the functions implemented in this PR

    mod_names = idx.get_module_names()
    if len(mod_names) > 1:
        raise ValueError("Provided YAML contains multiple modules {0}. "
                           "It is ambiguous which one to use."
                           .format(mod_names))

    mod_name = idx.get_module_names()[0]
    mod = idx.get_module(mod_name)
    streams = mod.get_all_streams()
    if len(streams) > 1:
        raise ValueError("Provided YAML contains multiple streams {0} of "
                           "a module '{1}'. It is ambiguous which one to use."
                           .format([x.get_stream_name() for x in streams],
                                   mod_name))

    mod_stream = streams[0]
    print(mod_stream.get_module_name())
    print(mod_stream.get_summary())
```

(This is a helper function `_yaml2stream` that I am proposing in https://github.com/rpm-software-management/modulemd-tools/pull/11/)

**Comment by FrostyX** [comment]
But it is quite possible, that I am talking nonsense here. I am not a `libmodulemd` developer, so I have a very limited understanding of how things work under the hood. In that case, please correct me.

**Comment by sgallagher** [comment]
> > I'm probably missing some context, but don't `modulemd_module_stream_read_file()` and `modulemd_module_stream_read_string()` provide that?
> 
> As far as I understand it, they don't. Those functions return `ModulemdModuleIndex` and there is AFAIK a cumbersome number of steps to actually access the attributes. Maybe I am doing it wrong, but didn't find any simpler way than this:

Yes, you are doing it wrong :sweat_smile: 

What @mmathesius was suggesting was that you can use [`Modulemd.ModuleStream.read_file()`](file:///home/sgallagh/workspace/libmodulemd/build/modulemd/html/ModulemdModuleStream.html#modulemd-module-stream-read-file) which returns a `Modulemd.ModuleStream` object. It will only work if the file or string you are reading contains *exactly* one YAML subdocument which is a ModuleStream, however. Otherwise, you must use the index.

**Comment by sgallagher** [comment]
I added the requested tests. They actually caused me to stumble across a few minor issues, so I put those in their own patches and pushed this along with it.

Also, Rawhide recently switched to python-black 20.8 which disagrees with python-black 19.10 about several formatting changes, so I've disabled the dirty-repo test, since both F33 or F34 will modify the code formatting differently (thus causing the test to fail on one or the other release).

File: modulemd/tests/test-version.sh [hunk 1]
```diff
@@ -30,7 +30,7 @@ OLDTAG=$(git describe --first-parent --abbrev=0)
 NEWVERSION=$(meson introspect $MESON_BUILD_ROOT --projectinfo |jq -r .version)
 NEWTAG=libmodulemd-$NEWVERSION
 
-if [ $NEWTAG = $OLDTAG ]; then
+if [ "$NEWTAG" = "$OLDTAG" ]; then
     error_out code=2 message="Version is already tagged. Update meson.build with the new version."
```
Comment (mmathesius) [inline modulemd/tests/test-version.sh:34]: Huh. What changed with newer bash versions? Something to do with the handling of empty strings?

File: modulemd/include/modulemd-2.0/modulemd-module-stream.h [hunk 2]
```diff
@@ -128,7 +128,7 @@ modulemd_module_stream_new (guint64 mdversion,
  * Returns: (transfer full): A newly-allocated #ModulemdModuleStream object if
  * the YAML file was valid and contained exactly one `document: modulemd`
  * subdocument. NULL if the document fails validation or multiple documents are
- * encountered and sets NULL appropriately. See #ModulemdModuleIndex for
+ * encountered and sets @error appropriately. See #ModulemdModuleIndex for
```
Comment (mmathesius) [inline modulemd/include/modulemd-2.0/modulemd-module-stream.h:131]: Good catches.

**Review by mmathesius — CHANGES_REQUESTED** [review]
Just a few nit picks.

File: modulemd/tests/test-version.sh (hunk 1 above)
Comment (sgallagher) [reply modulemd/tests/test-version.sh:34]: Yeah, it now requires quoting.

File: modulemd/tests/test-modulemd-common.c [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+main (int argc, char *argv[])
+{
+  setlocale (LC_ALL, "");
+
+  g_test_init (&argc, &argv, NULL);
+  g_test_bug_base ("https://bugzilla.redhat.com/show_bug.cgi?id=");
+
+  g_test_add_func ("/modulemd/v2/common/get_version",
+                   test_modulemd_get_version);
+  g_test_add_func ("/modulemd/v2/common/load_file", test_modulemd_load_file);
```
[the remark this answers is not in the document]
Comment (sgallagher) [reply modulemd/tests/test-modulemd-common.c:95]: Oof, I forgot to add one. I'll take you up on that offer.

File: modulemd/tests/test-modulemd-common.c [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  g_clear_error (&error);
+  g_clear_object (&idx);
+  g_clear_pointer (&yaml_file, g_free);
+  yaml_file = g_strdup_printf ("%s/nl.po", g_getenv ("TEST_DATA_PATH"));
+  idx = modulemd_load_file (yaml_file, &error);
+  g_assert_null (idx);
+  g_assert_error (error, MODULEMD_YAML_ERROR, MMD_YAML_ERROR_UNPARSEABLE);
+
+
+  /* Readable, non-YAML file */
```
[the remark this answers is not in the document]
Comment (sgallagher) [reply modulemd/tests/test-modulemd-common.c]: Fixed in the latest push.

**Review by mmathesius — APPROVED** [review]

**Pull request merged**